### ISSUE TITLE
A different assignment is trying to be approved (and failing) than the one I specify in the command line

### ISSUE THREAD

**Issue by ttomita** [NYUCCL/psiTurk#360]
I created a HIT in sandbox mode and completed the HIT myself. When I attempt to approve this particular assignment (3TXWC2NHN02DZ2CL7KUQSLQLTYB9S1), an error message is displayed saying that a different assignment (3NGMS9VZTMUDW7YC4M011EINXJMFFB) - one from a previous HIT and that has already been approved - does not exist (but it does exist):

```
[psiTurk server:off mode:sdbx #HITs:0]$ worker approve 3TXWC2NHN02DZ2CL7KUQSLQLTYB9S1
Approving specified submissions for the current study...
Response(Operation: approve_assignment_by_assignment_id | Status: error | Exception: An error occurred (RequestError) when calling the ApproveAssignment operation: Assignment 3NGMS9VZTMUDW7YC4M011EINXJMFFB does not exist. (1563504162075 s))
```

Oddly enough, when I try to reject assignment 3TXWC2NHN02DZ2CL7KUQSLQLTYB9S1, it works:

```
[psiTurk server:off mode:sdbx #HITs:0]$ worker reject 3TXWC2NHN02DZ2CL7KUQSLQLTYB9S1
Response(Operation: reject_assignment | Status: success)
```

Furthermore, now if I unreject the assignment and then try to reject it again, it says it can't because the operation can only be called with a status of Submitted:

```
[psiTurk server:off mode:sdbx #HITs:0]$ worker unreject 3TXWC2NHN02DZ2CL7KUQSLQLTYB9S1
Response(Operation: unreject_assignment | Status: success)
[psiTurk server:off mode:sdbx #HITs:0]$ worker reject 3TXWC2NHN02DZ2CL7KUQSLQLTYB9S1
Response(Operation: reject_assignment | Status: error | Exception: An error occurred (RequestError) when calling the RejectAssignment operation: This operation can be called with a status of: Submitted (1563505657987 s))
```

But shouldn't the status have reverted back to Submitted when it was unrejected?

**Comment by deargle** [comment]
I suspect that I didn't think through some repeat-worker-id-in-database
edge cases in the refactor. It's probably taking the assignmentid, looking
up the corresponding participant record in the local db, and then at some
point pulling that record's assignmentid, which doesn't match the first one
given. I'll take a look in a bit. Code gets hairy and i dont have a unit
test for that scenario yet.

Do you have allow_repeats allowed in your config? You must, right? Else you
wouldnt have been able to do the assignment twice? Or... did you wipe your
db?

**Comment by deargle** [comment]
Oh. and no, when you unreject an assignment, you have in effect approved
it, and you cannot undo that. That's an mturk thing -- not a psiturk thing.

**Comment by ttomita** [comment]
I have `allow_repeats = false`, and I wipe the db and restore from the previous state in order to do the assignment again.

I didn't know that was the behavior of MTurk's unreject (I've never actually had to use it for live HITs). Thanks for the tip!

I should also note that trying to reject all assignments in a HIT using `worker reject --hit <HIT ID>` has similar strange behavior to what I noted above. It also attempts to reject assignments not pertinent to the specified HIT and says those assignments do not exist.

**Comment by deargle** [comment]
Egads, found it. On it.

**Comment by deargle** [comment]
Just so I can confirm for testing, would you be open to sharing a dump of your db with me? dave@daveeargle.com

**Comment by deargle** [comment]
Actually, i don't understand the first mismatched-assignmentid one yet. But for the weird assignments in `worker reject --hit <hitid>`:

There is a bug in 2.3.2 right here: https://github.com/NYUCCL/psiTurk/blob/be169c67c2e1e1be05b5a0b9d5333803aff3d335/psiturk/amt_services_wrapper.py#L304

that line should be `query = query.filter(Participant.hitid == hit_id)` -- that explains the `worker reject --hit <hit_id>` problem you're seeing. I have it fixed on master right now. The consequence of that bug is that it doesn't respect the --hit specified and instead tries to do everything.

**Comment by deargle** [comment]
Actually, I have a tentative hypothesis for the mismatched one. But it's odd and probably not fully right. I found another bug in there -- that when trying to approve a specified hit, it looks for an _already approved_ hit in the local database to try to approve (whoops), when instead it should be looking for a _submitted_ status assignment.

https://github.com/NYUCCL/psiTurk/blob/be169c67c2e1e1be05b5a0b9d5333803aff3d335/psiturk/amt_services_wrapper.py#L240-L259

For this hypothesis to be right, line 249 in there would be to be being ignored, and then later 265 queries for the first matching record, which would give a record with the wrong assignment_id. Also, you would need to be trying to approve an already-approved assignmentid, which is possible given that you've been tinkering with unrejecting. That's a lot of holes to be aligning, but it's possible.

At any rate, I usually just stick with `worker approve --all`.

**Comment by ttomita** [comment]
I attempted to install from the master branch so that I could have the `worker reject --hit` fix you mentioned above. This required me to upgrade to python 3 because the updated `cmd2==0.9.13` requirement doesn't support python 2.7. I installed python 3.7.3 with pip in a clean conda virtual environment, followed by `sudo pip install git+git://github.com/NYUCCL/psiTurk.git@master`. Installation went smoothly with no warnings or errors. However, when I try to run any command on the psiTurk command line, now this happens:

```
[psiTurk server:off mode:sdbx #HITs:0]$ hit list
python(89444,0x1177455c0) malloc: *** error for object 0x10fe72d70: pointer being freed was not allocated
python(89444,0x1177455c0) malloc: *** set a breakpoint in malloc_error_break to debug
Abort trap: 6
```

The same thing happens when I pip install the latest official release too. This does not happen within my python 2 conda virtual environment, which has the latest official release installed.

As long as `worker reject <assignment_id>` and `worker approve --all` (haven't tried it yet) are functional, this might not be super urgent and I should be able to manage.
